File: javascript/test/instance/GenericCollectionHolder_GetOrElseAlias.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {IndexWithReturnCallback}         from "../../src/type/callback"
import type {StraightCollectionHolderForTest} from "./StraightCollectionHolderForTest"

import {GenericCollectionHolder} from "../../src/GenericCollectionHolder"
import {ABCD}                    from "../value/arrays"

export class GenericCollectionHolder_GetOrElseAlias
    extends GenericCollectionHolder<string>
    implements StraightCollectionHolderForTest<string> {

    public amountOfCall = 0

    public constructor() { super(ABCD,) }

    public execute(action: (instance: this,) => void,): this {
        action(this,)
        return this
    }

    public override getOrElse<const U, >(index: number, defaultValue: IndexWithReturnCallback<U>,): | string | U
    public override getOrElse(index: number, defaultValue: IndexWithReturnCallback<string>,): string
    public override getOrElse(index: number, defaultValue: IndexWithReturnCallback<unknown>,) {
        this.amountOfCall++
        return super.getOrElse(index, defaultValue,)
    }

}

File: java/collection.joookiwi/src/test/java/instance/GenericCollectionHolder_GetOrElseAlias.java
package instance;

import java.util.function.Consumer;
import java.util.function.IntFunction;
import java.util.function.Supplier;
import joookiwi.collection.java.GenericCollectionHolder;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_THIS_1;
import static joookiwi.collection.java.NumericConstants.MAX_INT_VALUE;
import static value.Arrays.ABCD;

@NotNullByDefault
public final class GenericCollectionHolder_GetOrElseAlias
        extends GenericCollectionHolder<@Nullable String>
        implements StraightCollectionHolderForTest<String, GenericCollectionHolder_GetOrElseAlias> {

    /// The field that tell the amount of time the methods `getOrElse` have been called
    public @Range(from = 0, to = MAX_INT_VALUE) int amountOfCall = 0;

    public GenericCollectionHolder_GetOrElseAlias() { super(ABCD); }

    @Override public @Range(from = 0, to = MAX_INT_VALUE) int getAmountOfCall() { return amountOfCall; }

    @Contract(ALWAYS_THIS_1)
    @Override public GenericCollectionHolder_GetOrElseAlias execute(final Consumer<GenericCollectionHolder_GetOrElseAlias> action) {
        action.accept(this);
        return this;
    }

    @Override public @Nullable String getOrElse(final int index, final IntFunction<? extends String> defaultValue) {
        amountOfCall++;
        return super.getOrElse(index, defaultValue);
    }

    @Override public @Nullable String getOrElse(final int index, final Supplier<? extends String> defaultValue) {
        amountOfCall++;
        return super.getOrElse(index, defaultValue);
    }

}
